src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -982,6 +982,257 @@ private Task WriteSuffix()
             Output.ProducingComplete(end);
         }
 
+        public bool TakeStartLineSpan(ReadableBuffer buffer, out ReadCursor consumed, out ReadCursor examined)
+        {
+            var start = buffer.Start;
+            var end = buffer.Start;
+
+            examined = buffer.End;
+            consumed = buffer.Start;
+
+            if (_requestProcessingStatus == RequestProcessingStatus.RequestPending)
+            {
+                ConnectionControl.ResetTimeout(_requestHeadersTimeoutMilliseconds, TimeoutAction.SendTimeoutResponse);
+            }
+
+            _requestProcessingStatus = RequestProcessingStatus.RequestStarted;
+
+            ReadableBuffer limitedBuffer;
+            if (buffer.Length >= ServerOptions.Limits.MaxRequestLineSize)
+            {
+                limitedBuffer = buffer.Slice(0, ServerOptions.Limits.MaxRequestLineSize);
+            }
+            else
+            {
+                limitedBuffer = buffer;
+            }
+
+            if (ReadCursorOperations.Seek(limitedBuffer.Start, limitedBuffer.End, out end, ByteLF) == -1)
+            {
+                if (limitedBuffer.Length == ServerOptions.Limits.MaxRequestLineSize)
+                {
+                    RejectRequest(RequestRejectionReason.RequestLineTooLong);
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            const int stackAllocLimit = 512;
+
+            // Move 1 byte past the \r
+            end = limitedBuffer.Move(end, 1);
+            var startLineBuffer = limitedBuffer.Slice(0, end);
+
+            Span<byte> span;
+
+            if (startLineBuffer.IsSingleSpan)
```

Comment:
I'm not making that change as part of this PR. I attempted this earlier and realized that the buffer size would be an issue. We need to have smart logic to use the cached buffer if it was <= the size of the buffers crossing the spans or new up one from scratch if > than the buffer size. I think we should look at that after this change goes in.